File: Main/ArchiMeter.Analysis/DependencyChain.cs
namespace ArchiMeter.Analysis
{
	using System;
	using System.Collections.Generic;
	using System.Linq;

	using ArchiMeter.Common;

	public class DependencyChain
	{
		private readonly EdgeItem[] _chain;

		public DependencyChain(IEnumerable<EdgeItem> referenceChain, EdgeItem root, EdgeItem lastEdge)
		{
			_chain = referenceChain.Concat(new[] { lastEdge }).ToArray();
			Length = _chain.Length;
			Root = root;
			LastEdge = lastEdge;
			IsCircular = root.Dependant == lastEdge.Dependency;
			Name = GetName();
		}

		public string Name { get; private set; }

		public IEnumerable<EdgeItem> ReferenceChain
		{
			get { return _chain; }
		}

		public EdgeItem Root { get; private set; }

		public EdgeItem LastEdge { get; private set; }

		public int Length { get; private set; }

		public bool IsCircular { get; private set; }

		public bool IsContinuation(EdgeItem edge)
		{
			return LastEdge.Dependency == edge.Dependant && !(ReferenceChain.Any(e => e.Dependant == edge.Dependant));
		}

		public bool Contains(EdgeItem edge)
		{
			return ReferenceChain.Any(e => e.Dependant == edge.Dependant && e.Dependency == edge.Dependency);
		}

		private string GetName()
		{
			var firstLink = ReferenceChain.Select(e => e.Dependant).OrderBy(x => x).FirstOrDefault();
			var startIndex = Array.FindIndex(_chain, e => e.Dependant == firstLink);
			var startSeq = ReferenceChain.Skip(startIndex);
			var endSeq = ReferenceChain.Take(startIndex);
			return string.Join(" --> ", startSeq.Concat(endSeq));
		}

		public override string ToString()
		{
			return Name;
		}

		public override int GetHashCode()
		{
			return Name.GetHashCode();
		}

		public override bool Equals(object obj)
		{
			return obj != null && obj.GetHashCode() == GetHashCode();
		}
	}
}

File: ArchiMeter.Analysis.FS/DependencyChain.fs
namespace ArchiMeter.Analysis
open ArchiMeter.Common
open System
open System.Collections.Generic

type DependencyChain(referenceChain : seq<EdgeItem>, root : EdgeItem, lastEdge : EdgeItem) = 
    let keys = fun (e : EdgeItem) -> e.Dependant
    let _rootDependant = root.Dependant
    let _lastDependency = lastEdge.Dependency
    let _referenceChain = Seq.concat [referenceChain; Seq.ofArray [|lastEdge|]] |> Seq.toArray
    let getName() = 
        let firstLink = _referenceChain |> Seq.map keys |> Seq.sort |> Seq.head
        let startIndex = _referenceChain |> Seq.findIndex (fun e -> e.Dependant = firstLink)
        let startSeq = _referenceChain |> Seq.skip startIndex
        let endSeq = _referenceChain |> Seq.take startIndex
        String.Join(" --> ", [|startSeq; endSeq|] |> Seq.concat)
    let _name = getName()
    let _hashCode = _name.GetHashCode()
    member this.ReferenceChain = _referenceChain |> Seq.ofArray
    member this.Root = root
    member this.LastEdge = lastEdge
    member this.ChainLength = Seq.length _referenceChain
    member this.IsCircular = _rootDependant = _lastDependency
    member this.IsContinuation(edge : EdgeItem) = _lastDependency = edge.Dependant && not (_referenceChain |> Seq.exists(fun (e : EdgeItem) -> e.Dependant = edge.Dependant))
    member this.Contains(edge : EdgeItem) = _referenceChain |> Seq.exists (fun e -> e.Dependant = edge.Dependant && e.Dependency = edge.Dependency)
    override this.ToString() = _name
    override this.GetHashCode() = _hashCode
    override this.Equals(o : obj) =
        match o with
        | null -> false
        | :? DependencyChain as d -> d.GetHashCode() = _hashCode
        | _ -> false
